File: services/reservation/src/main/java/com/example/iter/reservation/util/ReturnMapper.java
package com.example.iter.reservation.util;

import com.example.iter.auth.api.UserSummary;
import com.example.iter.reservation.domain.entity.Receipt;
import com.example.iter.reservation.domain.entity.Rental;
import com.example.iter.reservation.domain.entity.ReturnReceipt;
import com.example.iter.reservation.dto.response.ConditionEvidenceResponse;
import com.example.iter.reservation.dto.response.ReturnComparisonResponse;
import com.example.iter.reservation.dto.response.ReturnConfirmationResponse;
import com.example.iter.reservation.dto.response.ReturnTargetResponse;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class ReturnMapper {

    // 반납 확인 대상 목록 응답으로 변환합니다.
    public ReturnTargetResponse toTarget(
            Rental rental,
            UserSummary renter,
            String thumbnailUrl,
            ReturnReceipt returnReceipt
    ) {
        return new ReturnTargetResponse(
                rental.getId(),
                rental.getProductNameSnapshot(),
                thumbnailUrl,
                renter,
                rental.getEndDate(),
                returnReceipt.getReturnDate()
        );
    }

    // 수령·반납 증빙 비교 응답으로 변환합니다.
    public ReturnComparisonResponse toComparison(
            Rental rental,
            UserSummary renter,
            Receipt receipt,
            List<String> receiptImageUrls,
            ReturnReceipt returnReceipt,
            List<String> returnImageUrls
    ) {
        return new ReturnComparisonResponse(
                rental.getId(),
                rental.getProductNameSnapshot(),
                renter,
                rental.getStartDate(),
                rental.getEndDate(),
                returnReceipt.getReturnDate(),
                toEvidence(receipt, receiptImageUrls),
                toEvidence(returnReceipt, returnImageUrls)
        );
    }

    // 반납 최종 확인 응답으로 변환합니다.
    public ReturnConfirmationResponse toConfirmation(
            Rental rental,
            Long disputeId
    ) {
        return new ReturnConfirmationResponse(
                rental.getId(),
                rental.getStatus(),
                disputeId
        );
    }

    // ============================================================

    private ConditionEvidenceResponse toEvidence(
            Receipt receipt,
            List<String> imageUrls
    ) {
        return new ConditionEvidenceResponse(
                receipt.getProductCondition(),
                receipt.getConditionDetail(),
                imageUrls,
                receipt.getReceivedAt()
        );
    }

    private ConditionEvidenceResponse toEvidence(
            ReturnReceipt returnReceipt,
            List<String> imageUrls
    ) {
        return new ConditionEvidenceResponse(
                returnReceipt.getProductCondition(),
                returnReceipt.getConditionDetail(),
                imageUrls,
                returnReceipt.getCreatedAt()
        );
    }

}

File: services/reservation/src/main/kotlin/com/example/iter/reservation/util/ReturnMapper.kt
package com.example.iter.reservation.util

import com.example.iter.auth.api.UserSummary
import com.example.iter.reservation.domain.entity.Receipt
import com.example.iter.reservation.domain.entity.Rental
import com.example.iter.reservation.domain.entity.ReturnReceipt
import com.example.iter.reservation.dto.response.ConditionEvidenceImageResponse
import com.example.iter.reservation.dto.response.ConditionEvidenceResponse
import com.example.iter.reservation.dto.response.ReturnComparisonResponse
import com.example.iter.reservation.dto.response.ReturnConfirmationResponse
import com.example.iter.reservation.dto.response.ReturnTargetResponse
import org.springframework.stereotype.Component

@Component
class ReturnMapper {

    fun toTarget(
        rental: Rental,
        renter: UserSummary,
        thumbnailUrl: String?,
        returnReceipt: ReturnReceipt
    ): ReturnTargetResponse = ReturnTargetResponse(
        rental.id,
        rental.productNameSnapshot,
        thumbnailUrl,
        renter,
        rental.endDate,
        returnReceipt.returnDate
    )

    fun toComparison(
        rental: Rental,
        renter: UserSummary,
        listingImages: List<ConditionEvidenceImageResponse>,
        receipt: Receipt,
        receiptImages: List<ConditionEvidenceImageResponse>,
        returnReceipt: ReturnReceipt,
        returnImages: List<ConditionEvidenceImageResponse>
    ): ReturnComparisonResponse = ReturnComparisonResponse(
        rental.id,
        rental.productNameSnapshot,
        renter,
        rental.startDate,
        rental.endDate,
        returnReceipt.returnDate,
        listingImages.toList(),
        toEvidence(receipt, receiptImages),
        toEvidence(returnReceipt, returnImages)
    )

    fun toConfirmation(rental: Rental, disputeId: Long?, reportId: Long?): ReturnConfirmationResponse = ReturnConfirmationResponse(
        rental.id,
        rental.status,
        disputeId,
        reportId
    )

    private fun toEvidence(receipt: Receipt, images: List<ConditionEvidenceImageResponse>): ConditionEvidenceResponse = ConditionEvidenceResponse(
        receipt.productCondition,
        receipt.conditionDetail,
        images.toList(),
        receipt.receivedAt
    )

    private fun toEvidence(
        returnReceipt: ReturnReceipt,
        images: List<ConditionEvidenceImageResponse>
    ): ConditionEvidenceResponse = ConditionEvidenceResponse(
        returnReceipt.productCondition,
        returnReceipt.conditionDetail,
        images.toList(),
        returnReceipt.createdAt
    )
}
